accept-guard.c

```diff
@@ -292,19 +307,29 @@ static int peek_ifindex(int sd)
 	struct sockaddr_in sin;
 	struct cmsghdr *cmsg;
 	struct msghdr msgh;
+	socklen_t orig_len;
+	int orig_on = 0;
 	int on = 1;
 
-	setsockopt(sd, SOL_IP, IP_PKTINFO, &on, sizeof(on));
+	orig_len = sizeof(orig_on);
+	if (getsockopt(sd, SOL_IP, IP_PKTINFO, &orig_on, &orig_len) == -1)
+		return 0;	/* Fall back to allow syscall on error */
+	if (setsockopt(sd, SOL_IP, IP_PKTINFO, &on, sizeof(on)) == -1)
+		return 0;	/* Fall back to allow syscall on error */
 
 	memset(&msgh, 0, sizeof(msgh));
 	msgh.msg_name = &sin;
 	msgh.msg_namelen = sizeof(sin);
 	msgh.msg_control = cmbuf;
 	msgh.msg_controllen = sizeof(cmbuf);
 
-	if (org_recvmsg(sd, &msgh, MSG_PEEK) == -1)
+
+	if (org_recvmsg(sd, &msgh, MSG_PEEK) == -1) {
+		setsockopt(sd, SOL_IP, IP_PKTINFO, &orig_on, sizeof(orig_on));
```

Comment:
I can't think of anything worthwhile to do here if `setsockopt` return `-1` (I guess it could happen if the socket has disappeared), but please let me know if you can think of anything!

Reply:
If either of these two fail there's not much we can do.  The best we can do is to mark it as "ignore" the old-school way for posterity.  I.e.,

```C
    (void)setsockopt(...);
```